gui/src/ruler.cc

```diff
@@ -0,0 +1,362 @@
+    gtk_widget_queue_draw(drawingArea);
+}
+
+void Ruler::update()
+{
+    if (drawingArea == nullptr) { return; }
+
+    // We need to calculate the distance between the largest ticks on the ruler
+    // We will try each interval sequentially until we find an interval which
+    // will produce segments of a large enough width/height when drawn
+
+    // Once we have the distance between the major ticks (the size of the segments)
+    // each segment is then divided into 5 parts assuming there's enough space
+    // and each of those parts is again divided into 2, again, assuming there's enough space
+
+    // Index in the ruler's VALID_INTERVALS array
+    int intervalIndex = 0;
+    // Each interval is multiplied by 10 raised to the power n
+    const int INTERVAL_BASE = 10;
+    int intervalN = 0;
+
+    while (true)
+    {
+        if (VALID_INTERVALS.at(intervalIndex) != 1 || intervalN != 0)
+        {
+            majorInterval = VALID_INTERVALS.at(intervalIndex) * pow(INTERVAL_BASE, intervalN);
+
+            // We check versus the width or height depending on orientation
+            double rulerSize = NAN;
+            if (orientation == HORIZONTAL)
+            {
+                rulerSize = width;
+            }
+            else
+            {
+                rulerSize = height;
+            }
+            // Calculate the drawn size for this interval by mapping from the ruler range
+            // to the ruler size on the screen
+            segmentScreenSize = floor(mapRange(majorInterval, 0.0, upperLimit - lowerLimit, 0.0, rulerSize));
+
+            // If we've found a segment of appropriate size, we can stop
+            if (segmentScreenSize >= MIN_SEGMENT_SIZE) { break; }
+        }
+
+        // Try the next interval
+        intervalIndex++;
+        if (intervalIndex == VALID_INTERVALS.size())
+        {
+            // We tried all intervals for the current n
+            intervalIndex = 0;
+            intervalN++;
+        }
+    }
+}
+
+double Ruler::mapRange(double x, double a_lower, double a_upper, double b_lower, double b_upper)
+{
+    double a_width = a_upper - a_lower;
+    double b_width = b_upper - b_lower;
+    double scale = b_width / a_width;
+
+    return b_lower + scale * (x - a_lower);
+}
+
+gboolean Ruler::drawCallback(GtkWidget *widget, cairo_t *cr, gpointer data)
+{
+
+    auto *ruler = static_cast<Ruler *>(data);
+
+    if (ruler->drawingArea == nullptr) { return FALSE; }
+
+    double width = ruler->width;
+    double height = ruler->height;
+
+    // Draw background
+    GtkStyleContext *context = gtk_widget_get_style_context(widget);
+    gtk_render_background(context, cr, 0, 0, ruler->width, ruler->height);
+
+    // Draw outline along left and right sides and along the bottom
+    gdk_cairo_set_source_rgba(cr, &(ruler->lineColor));
+
+    cairo_set_line_width(cr, Ruler::LINE_WIDTH);
+    if (ruler->orientation == HORIZONTAL)
+    {
+        cairo_move_to(cr, LINE_COORD_OFFSET, 0);
```

Comment:
Maybe for each of these three blocks a comment to indicate what you are drawing? Bonus points if you stuff each of these blocks into a function, called `drawWhateverItIsYouAreDrawing()` ;-)